core/domain/exp_domain.py

```diff
@@ -1978,28 +1951,32 @@ def validate(self, strict=False):
         for state_name in self.states:
             self._require_valid_state_name(state_name)
             state = self.states[state_name]
-            self.states[state_name].validate(
+            state.validate(
+                self.param_specs,
                 allow_null_interaction=not strict)
             # The tests below are shifted from validate() in class
-            # InteractionInstance as these tests are exclusive to an interaction
-            # in a state in an exploration, and not for that in a question,
-            # (whereas the validate() in InteractionInstance is also used by
-            # the question_data field).
-            if state.interaction.id is not None:
-                if not isinstance(state.interaction.id, basestring):
+            # Outcome (which are called from the validate() function in class
```

Comment:
Don't talk about "shifting" in the docstring; new devs won't know what that refers to since they only see the new version.

Something like "The checks below perform validation on the Outcome domain object that is specific to answer groups in explorations, but not questions. This logic is here because the validation checks in the Outcome domain object are used by both explorations and questions." should suffice.

Reply:
Done.